### DISCUSSION TITLE
a few more performance tweaks

### DISCUSSION THREAD

**Discussion by frankozland** [RaspAP/raspap-webgui#1651] [Ideas]
ootb rasperry wont load balance irqs properly across cpu's
The fix is manually set affinity - i tried installing irqbalance,but that didnt do anything.

some manual tweeks that added another 3-6mbps....this is with raspap + wireguard, bbr, affinity tweaks. Notice even distribution across cpu's:

![image](https://github.com/user-attachments/assets/a9e2e3f1-4a6f-4f10-8f82-387fad423327)

https://tomcore.io/docs/articles/RaspberryPI/raspberry-pi-4b-network-performance/
```
cat /proc/interrupts
#eth0 is pinned to cpu 0 - this means this 1 cpu is doing all the work; thats probably less than ideal.

#Better to move - problem is eth0 may show up on different irq's so you have to find which ones.
#To set new affinity once you have irq is:
echo 2 > /proc/irq/38/smp_affinity
echo 4 > /proc/irq/39/smp_affinity

#find mmc - move that off cpu0.  MMC is wifi
echo a > /proc/irq/37/smp_affinity



echo f >/sys/class/net/eth0/queues/tx-0/xps_cpus
echo f >/sys/class/net/eth0/queues/tx-1/xps_cpus
echo f >/sys/class/net/eth0/queues/tx-2/xps_cpus
echo f >/sys/class/net/eth0/queues/tx-3/xps_cpus
echo f >/sys/class/net/eth0/queues/tx-4/xps_cpus
echo f >/sys/class/net/eth0/queues/rx-0/rps_cpus
```

And I added all of this to sysctl.conf and then sysctl -p
https://thelinuxnotes.com/index.php/how-to-tweak-sysctl-for-high-load/

I was able to push 60mbps from vnam to georgia over wireguard which i think is nothing short of amazing.
This is very close to actual speed to georgia w/o raspap:
![image](https://github.com/user-attachments/assets/848e495f-102c-4b20-a54e-a885e31d23af)

iftop results
![image](https://github.com/user-attachments/assets/ab3f9ac6-eb07-4da0-a097-7657c981420f)

irq's are a bit better balanced
![image](https://github.com/user-attachments/assets/d83631e4-b6af-4606-b06d-8a4033cb17f5)

without these tweaks both mmc (the wifi interface) and eth0 are concentrated on cpu0 and not spread across cpu's
RaspAP is not optimized as CPU0 is doing all the work.

The nice to have would be to tweak configuration on install to optimal performance for heavy users.
Possibly a 2nd script for high optimize settings for insiders?

![image](https://github.com/user-attachments/assets/3eefd8b5-0c3f-45b4-8410-410f001a9093)

**Comment by frankozland** [comment]
Throw some traffic shaping at it - eth0 will default to qdisc mq even with BBR enabled.

So, we can alter that.
```
tc qdisc replace dev eth0 root cake oceanic ack-filter bandwidth 100Mbps diffserv3 flows
```

CAKE  (Common  Applications Kept Enhanced) is a shaping-capable queue discipline which uses both AQM and FQ.
![image](https://github.com/user-attachments/assets/479d134e-4ea7-4892-910e-b119f58772a5)

What im asking this to do is change the queuing algorithm to use cake, im very remote so shape traffic with hi latency (oceanic) - my bandwidth is limited.  I am being aggressive and filtering out acks where i can. And provide a reasonable queuing algorithm with diffserv3 and flows.

Because im so remote, "oceanic" should smooth out some the latency issues.

Heres some documentation on tc....https://man7.org/linux/man-pages/man8/tc-cake.8.html

I think this setting has to be set on every boot to be in effect. And i think it requires host support bbr.

http://arxiv.org/pdf/1804.07617
https://lwn.net/Articles/758353/
![image](https://github.com/user-attachments/assets/f1ff0e95-0e9e-4e8c-9ceb-1c8bd6b90e01)


To see the results:

```
$ tc -s qdisc show dev eth0
qdisc cake 8002: root refcnt 6 bandwidth 800Mbit diffserv3 flows nonat nowash ack-filter no-split-gso rtt 300ms noatm overhead 38 mpu 84 
 Sent 1071187091 bytes 1182280 pkt (dropped 1, overlimits 60179 requeues 3) 
 backlog 0b 0p requeues 3
 memory used: 18176b of 15140Kb
 capacity estimate: 8Gbit
 min/max network layer size:           28 /    1468
 min/max overhead-adjusted size:       84 /    1506
 average network hdr offset:           14

                   Bulk  Best Effort        Voice
  thresh         50Mbit      800Mbit      200Mbit
  target           15ms         15ms         15ms
  interval        300ms        300ms        300ms
  pk_delay          0us          6us         11us
  av_delay          0us          3us          1us
  sp_delay          0us          2us          1us
  backlog            0b           0b           0b
  pkts                0      1182227           54
  bytes               0   1071184977         2268
  way_inds            0          179            0
  way_miss            0          185            1
  way_cols            0            0            0
  drops               0            1            0
  marks               0            0            0
  ack_drop            0            0            0
  sp_flows            0            3            1
  bk_flows            0            1            0
  un_flows            0            0            0
  max_len             0         1482           42
  quantum          1514         1514         1514
```

**Comment by frankozland** [comment]
https://cs.uwaterloo.ca/~brecht/servers/apic/SMP-affinity.txt

**Comment by ikem-krueger** [comment]
Do you know similar tweaks for the Raspberry Pi 3B/3B+?
